core/trino-main/src/main/java/io/trino/server/protocol/Query.java

```diff
@@ -348,10 +347,15 @@ public synchronized ListenableFuture<QueryResults> waitForResults(long token, Ur
         return Futures.transform(futureStateChange, ignored -> getNextResult(token, uriInfo, targetResultSize), resultsProcessorExecutor);
     }
 
-    public synchronized void markResultsConsumedIfReady()
+    public void markResultsConsumedIfReady()
     {
-        if (!resultsConsumed && exchangeDataSource.isFinished()) {
-            queryManager.resultsConsumed(queryId);
+        if (resultsConsumed) {
+            return;
+        }
+        synchronized (this) {
+            if (!resultsConsumed && exchangeDataSource.isFinished()) {
+                queryManager.resultsConsumed(queryId);
+            }
         }
     }
 
```

Comment:
In `closeExchangeIfNecessary`, could we do the below before taking lock ?
```
if (queryInfo.getState() != FAILED && queryInfo.getOutputStage().isPresent()) {
            return;
}
```

Reply:
I saw `closeExchangeIfNecessary` from this class show up a lot in JFR, the above could be a way to reduce locking.